Jenkinsfile: migration to shared library & cleanup

peat-psuwit opened the pull request:
This PR moves all code for Jenkinsfile to a shared library that can be included from all repositories. This makes sure we DRY and allows us to implement a fix that applies to all repositories (should there be a need in the future).

CI needs to include this repository in the library list to be able to utilize the shared library.

peat-psuwit committed:
Jenkinsfile: migrate to shared-library-based pipeline

This commit moves all logic from the (example) Jenkinsfile to a shared
library file that can be loaded from multiple repositories. This
requires adding this repository into Jenkins server's shared library
list.

This is simply a copy-and-paste verbatim. More cleanup is comming in the
following commits.

peat-psuwit committed:
BaPDP: use only single line for build-binary invocation

There's no need for a separated export line; you can put the environment
variable in front of the command.

peat-psuwit committed:
BaPDP: remove extra word from top level build-binary stage

peat-psuwit committed:
BaPDP: migrate to declarative parallel syntax

peat-psuwit committed:
BaPDP: run AMD64 and ARMHF only for arch-dependent packages

mariogrip commented on vars/buildAndProvideDebianPackage.groovy:
```diff
@@ -0,0 +1,69 @@
+          }
+          stage('Build binary - arm64') {
+            agent { label 'arm64' }
+            // Always run; arch-independent packages are built here.
```
We might want to do arch independent packages on amd64 as its much faster

We can achieve this by setting
```
export architecture="all"
export FORCE_BINARY_ONLY="all"
```

peat-psuwit replied:
Maybe it's better to set `MAIN_ARCHITECTURE` correctly in `/etc/jenkins/debian_glue`? This way, packages with a mix of arch-dependent & arch-independent binary packages will benefit, too.

The condition & missing of it has to match the architecture that will build arch-independent packages.

mariogrip replied:
This as been done, "all" is now built on amd64

peat-psuwit committed:
BaPDP: move file list into method scope

Seems like code in method scope can't find variables declare at the top
level. I suspect they become (kind of) class statics?

peat-psuwit commented:
A build using `buildAndProvideDebianPackage()`.
https://ci.ubports.com/blue/organizations/jenkins/ubports%2Fgst-droid-packaging/detail/xenial_-_gst-droid/7/pipeline

A build using `buildAndProvideDebianPackage(/* isArchIndependent */ true)` (seems like BlueOcean's live view is confused by `when` directive, but the build is fine it seems).
https://ci.ubports.com/blue/organizations/jenkins/ubports%2Fdroidmedia-packaging/detail/xenial_-_gst-droid/8

peat-psuwit committed:
BaPDP: unstash armhf & amd64 only if !isArchIndependent

They're not run in arch-dependent only build, and thus can't be
unstashed.

peat-psuwit committed:
BaPDP: enable 'skipDefaultCheckout' option

Only 'Build source' stage requires checkout. Using this option allows us
to delete the initial cleanWs step at every stage but the first.

peat-psuwit committed:
Jenkinsfile: document the isArchIndependent parameter

peat-psuwit committed:
BaPDP: use checkoutToSubdirectory

This enables us to not moving to a subdirectory at the start of build-
source.sh (disabled with an env var).

peat-psuwit committed:
BaPDP: Move building of arch-independent package to AMD64

@mariogrip has updated CI's main architecture to AMD64. Thus, we have to
update this optimization too.

mariogrip reviewed (approved):
LGTM :+1:

UniversalSuperBox committed:
Add trailing newlines
